### ISSUE TITLE
API: Cannot Replace AAAA with CNAME in Single Request

### ISSUE THREAD

**Issue by domenukk** [desec-io/desec-stack#220]
Removing AAAA and adding a CNAME in the same request fails:
```python
PUT @ route /domains/enowars.com/rrsets/ [{'type': 'CNAME', 'records': ['elk.enowars.com.'], 'subname': 'kibana', 'ttl': 600}, {'type': 'AAAA', 'records': [], 'subname': 'kibana', 'ttl': 1}]
Request failed!
Unexpected error!
{"detail":"RRset kibana.enowars.com. IN CNAME: Conflicts with pre-existing non-CNAME RRset"}
```

The current entries were:
```bash
hnj@prokrastinator:~/Projects/ENOWARS2019/enonetwork/code/ansible$ dig ANY kibana.enowars.com

; <<>> DiG 9.10.3-P4-Ubuntu <<>> ANY kibana.enowars.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 43922
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;kibana.enowars.com.		IN	ANY

;; ANSWER SECTION:
kibana.enowars.com.	589	IN	RRSIG	AAAA 13 3 600 20190711000000 20190620000000 5301 enowars.com. Tf2LVZjOnrYj9myLmGgzLoz1+oM5HK2+33PM3lKUbEzv7SgS/4CfogIu 7AU1SaH1s8clbSlMT+ocV2vqLHBBRw==
kibana.enowars.com.	589	IN	AAAA	fd00:1337:0:1:e61f:13ff:fecb:3c5a

;; Query time: 4 msec
;; SERVER: 127.0.1.1#53(127.0.1.1)
;; WHEN: Tue Jul 02 17:50:31 CEST 2019
;; MSG SIZE  rcvd: 182
```

**Comment by nils-wisiol** [comment]
My guess is that the change tracker should either list the deletions first, or make to requests to pdns to avoid the conflict. Currently, [deletions are last in the list](https://github.com/desec-io/desec-stack/blob/f0d288b205e82eb5555e96f913424b8223ead827/api/desecapi/pdns_change_tracker.py#L127-L158).

**Comment by peterthomassen** [comment]
Let's create a test case for this.

May be related to https://github.com/PowerDNS/pdns/pull/7787 or https://github.com/PowerDNS/pdns/pull/6647 (and the underlying issues linked in the PRs). I'm reaching out to pdns to figure out if this is fixed in upcoming pdns 4.2.0 or not.

**Comment by peterthomassen** [comment]
Please see https://github.com/PowerDNS/pdns/issues/7501 and the following IRC excerpt from today:

> (10:40:09 PM) pt01: Habbie: alright, cool. We'll try the other way around, and my  understanding is that with what's known currently, it should work. Option 2 (large REPLACE to delete with records=[] and add the CNAME, regardless of order) should also work. Right?
> (10:40:25 PM) Habbie: pt01, i have no predictions on option 2
> (10:40:31 PM) Habbie: pt01, only the assertion that if it does not work, it is a bug

Together with the linked issue, it looks uncertain whether/when pdns will fix this. I suppose that switching the order is the easiest fix then.

**Comment by nils-wisiol** [comment]
This is surprisingly hard to fix. I formulated two requirements in b3028aa90386859b86c512d9d38321ad0521ccd9.

1. Do CNAME-update as expected when the other RRset is requrested to be deleted.
2. Don't do anything if the CNAME record contents fail validation.

The implementation is problematic, as validation relies on successful processing of pdns API requests.

- simply changing the order of entries in the single pdns-api request did not help (cf 706599cc792ef67685c9312f368f4282fa47730d)
- doing two requests to pdns, first deletion, second creation, will leave the transaction in a broken state if the second request fails (cf eb9ca8a594236481f8c1b8a35c6c09de65d98b09).

IMHO, this is a good reason to move validation from pdns to api and treat pdns-api 422 errors as api server errors that need the admin's attention.

For the short term, I suggest to not fix this issue.

For the long term, I'm voting to move validation to our API as outlined above.

**Comment by peterthomassen** [comment]
As an alternative, the `DELETE` could be incorporated in the `REPLACE` section (with empty `records` field). The "old code"™ did it like this, and if I'm not mistaken, I verified back then that this case worked (but not completely sure).

I'm not aware of any recommendation from pdns that favors `DELETE` over `REPLACE`.

**Comment by nils-wisiol** [comment]
After reading @Habbie's comment again, I changed the delete requests to use `REPLACE` with empty record set instead. This only works when the delete requests are specified first, see 62027c207afee4220f445ea094c20f8435d33765 and e75a3afc1655c4eeb80008c6e6331954e72a6c58, respectively.

**Comment by nils-wisiol** [comment]
I tested the four commits above again with pdns 4.2-rc2 and organized the results in a little table.

order | #requests | deletion method | commit | test result pdns 4.1 | test result pdns 4.2-rc2
------- | ------------- | -------------------- | ---------- | ----------------------- | --------------------------
delete then replace | 1 | by replace with [] | e75a3afc1655c4eeb80008c6e6331954e72a6c58 | ✔️ | ✔️ 
replace then delete | 1 | by replace with [] | 62027c207afee4220f445ea094c20f8435d33765 | status 422 | status 422
delete then replace | 1 | `DELETE` | 706599cc792ef67685c9312f368f4282fa47730d | status 422 | ✔️
replace then delete | 1 | `DELETE` | 2c20d11f895aa0e87f41e941edec7128c3e206b3 | status 422 | ✔️

Independent verification is very much appreciated!

**Comment by nils-wisiol** [comment]
The bottom-right result could be the "recommendation" for `DELETE` over empty `REPLACE` 🤷

**Comment by peterthomassen** [comment]
> (07:11:55 PM) Habbie: 4.2-rc2 row 4 is a surprise after row 2
> (07:12:06 PM) Habbie: as if we always handle the DELETEs first
> (07:14:15 PM) Habbie: i think us requiring an ordering is fine; which means i only demand that 1st and 3rd work
> (07:14:32 PM) Habbie: but 2nd failing while 4th work is a weird inconsistency, that warrants a ticket
> (07:14:58 PM) pt01: ok. So the canonical way to do it in 4.1 is option 1, and in 4.2 it is option 3. (Option 1 requires a dummy TTL, so option 3 is nicer)
> (07:15:06 PM) pt01: I will verify 2 vs. 4 and file a ticket if inconsistent. thx

**Comment by peterthomassen** [comment]
@nils-wisiol I reran cases 2 and 4. I can confirm 2, but I found that 4 fails. It turns out that your commit 2c20d11f895aa0e87f41e941edec7128c3e206b3 (case 4) was missing the new test case.

tl;dr pdns 4.2.0-rc2 allows replacing a record with a `CNAME` record *iff* the request contains the deletion before the addition, regardless of the deletion is done through `changetype="DELETE"` with `records=[]`, or `changetype="REPLACE"`. Updated table below (other cases non re-verified).

order | #requests | changetype for deletion | commit | test result pdns 4.1 | test result pdns 4.2-rc2
------- | ------------- | -------------------- | ---------- | ----------------------- | --------------------------
delete then replace | 1 | by replace with [] | e75a3afc1655c4eeb80008c6e6331954e72a6c58 | ✔️ | ✔️ 
replace then delete | 1 | by replace with [] | 62027c207afee4220f445ea094c20f8435d33765 | status 422 | status 422
delete then replace | 1 | `DELETE` | 706599cc792ef67685c9312f368f4282fa47730d | status 422 | ✔️
replace then delete | 1 | `DELETE` | 2c20d11f895aa0e87f41e941edec7128c3e206b3 | status 422 | status 422

@Habbie fyi. IIRC this is what you expected. so all good.

**Comment by Habbie** [comment]
> @Habbie fyi. IIRC this is what you expected. so all good.

Normality restored!
